library/cartridge_set_control_instance.py

```diff
@@ -15,14 +15,69 @@
 }
 
 
+GET_TWOPHASE_COMMIT_VERSION_TIMEOUT = 10
+
+
+def get_twophase_commit_versions(control_console, advertise_uris):
+    versions, err = control_console.eval_res_err('''
+        local fiber_clock = require('fiber').clock
+        local pool = require('cartridge.pool')
+
+        local uris, timeout = ...
+        local futures = {}
+
+        for i, uri in ipairs(uris) do
+            local conn, err = pool.connect(uri)
+            if err ~= nil then
+                return nil, err
+            end
+
+            local eval_opts = {
+                is_async = true,
+            }
+            local future = conn:eval([[
+                if rawget(_G, '__cartridge_upload_begin') ~= nil then
+                    return 2
+                else
+                    return 1
+                end
+            ]], nil, eval_opts)
+
+            futures[i] = future
+        end
+
+        local versions = {}
+
+        local deadline = fiber_clock() + timeout
+        for i, future in pairs(futures) do
+            local wait_timeout = deadline - fiber_clock()
+            if wait_timeout < 0 then
+                wait_timeout = 0
```

Comment:
return nil, 'timeout'?

Reply:
In this case you could get two different errors for timeout.

`nil, timeout` - at this line  and `nil, box.error.new(box.error.TIMEOUT)` few lines below.

It's only reasonable here to check that you correctly handle box.errors on Python side.

Reply:
I've added `tostring`.